src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java

```diff
@@ -0,0 +1,509 @@
+import org.apache.sling.auth.core.spi.AuthenticationInfo;
+import org.apache.sling.auth.core.spi.DefaultAuthenticationFeedbackHandler;
+import org.apache.sling.auth.oauth_client.ClientConnection;
+import org.apache.sling.auth.oauth_client.spi.LoginCookieManager;
+import org.apache.sling.auth.oauth_client.spi.OidcAuthCredentials;
+import org.apache.sling.auth.oauth_client.spi.UserInfoProcessor;
+import org.apache.sling.jcr.api.SlingRepository;
+import org.apache.sling.jcr.resource.api.JcrResourceConstants;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+import org.osgi.service.metatype.annotations.AttributeDefinition;
+import org.osgi.service.metatype.annotations.Designate;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Component(
+        service = AuthenticationHandler.class,
+        immediate = true
+)
+
+@Designate(ocd = OidcAuthenticationHandler.Config.class, factory = true)
+public class OidcAuthenticationHandler extends DefaultAuthenticationFeedbackHandler implements AuthenticationHandler {
+
+
+    private static final Logger logger = LoggerFactory.getLogger(OidcAuthenticationHandler.class);
+    private static final String AUTH_TYPE = "oidc";
+    public static final String REDIRECT_ATTRIBUTE_NAME = "sling.redirect";
+
+    private final SlingRepository repository;
+
+    private final Map<String, ClientConnection> connections;
+    private final OAuthStateManager stateManager;
+
+    private String idp;
+
+    private  final String callbackUri;
+
+    private LoginCookieManager loginCookieManager;
+
+    private String defaultRedirect;
+
+    private String defaultConnectionName;
+
+    private UserInfoProcessor userInfoProcessor;
+
+    private boolean userInfoEnabled;
+
+    // We don't want leave the cookie lying around for a long time because it it not needed.
+    // At the same time, some OAuth user authentication flows take a long time due to
+    // consent, account selection, 2FA, etc so we cannot make this too short.
+    protected static final int COOKIE_MAX_AGE_SECONDS = 300;
```

Comment:
ideally the constants should be private and if the default constants needs to be readable/editable -> introduce a public configuration option for the service.

Reply:
i think i made it private... hope that's ok